generator/src/main/java/com/scottlogic/deg/generator/walker/RandomReductiveDecisionTreeWalker.java

```diff
@@ -0,0 +1,34 @@
+package com.scottlogic.deg.generator.walker;
+
+import com.google.inject.Inject;
+import com.scottlogic.deg.generator.decisiontree.DecisionTree;
+import com.scottlogic.deg.generator.fieldspecs.RowSpec;
+
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class RandomReductiveDecisionTreeWalker implements DecisionTreeWalker {
+    private final ReductiveDecisionTreeWalker underlyingWalker;
+
+    @Inject
+    public RandomReductiveDecisionTreeWalker(ReductiveDecisionTreeWalker underlyingWalker) {
+        this.underlyingWalker = underlyingWalker;
+    }
+
+    @Override
+    public Stream<RowSpec> walk(DecisionTree tree) {
```

Comment:
Can this be done instead with 

    Stream.generate(()->getFirstResult(tree))
        .limit(maxRows);


    private RowSpec getFirstResult(DecisionTree tree) { 
        return underlyingWalker.walk(tree).findFirst().orElse(null);
    }

Reply:
Possibly, but we'd still need something to a) restart this generation and b) ensure it exits if there are no entries from the `Supplier`, which the `SourceRepeatingIterator` manages at the moment.

Certainly will remember `Stream.generate()` but not sure it will help _too_ much here.

Reply:
Refactored into a solution that used Stream.generate(), not lazy for the first value, but lazy afterwards and a 'simpler' solution.